llama-index-integrations/llms/llama-index-llms-gigachat/llama_index/llms/gigachat/base.py

```diff
@@ -0,0 +1,223 @@
+
+    @property
+    def metadata(self) -> LLMMetadata:
+        """Get LLM metadata."""
+        return LLMMetadata(
+            context_window=self.context_window,
+            num_output=self.context_window,
+            model_name=self.model,
+        )
+
+    @llm_completion_callback()
+    async def acomplete(
+        self,
+        prompt: str,
+        formatted: bool = False,
+        **kwargs: Any,
+    ) -> CompletionResponse:
+        """Get completion asynchronously."""
+        async with GigaChat(**self._gigachat_kwargs) as giga:
+            response = await giga.achat(
+                Chat(
+                    model=self.model,
+                    messages=[Messages(role="user", content=prompt)],
+                )
+            )
+        return CompletionResponse(
+            text=response.choices[0].message.content,
+        )
+
+    @llm_completion_callback()
+    def complete(
+        self,
+        prompt: str,
+        formatted: bool = False,
+        **kwargs: Any,
+    ) -> CompletionResponse:
+        """Get completion."""
+        with GigaChat(**self._gigachat_kwargs) as giga:
+            response = giga.chat(
+                Chat(
+                    model=self.model,
+                    messages=[Messages(role="user", content=prompt)],
+                )
+            )
+        return CompletionResponse(
+            text=response.choices[0].message.content,
+        )
+
+    @llm_chat_callback()
+    async def achat(
+        self,
+        messages: Sequence[ChatMessage],
+        **kwargs: Any,
+    ) -> ChatResponse:
+        """Get chat asynchronously."""
+        async with GigaChat(**self._gigachat_kwargs) as giga:
+            response = await giga.achat(
+                Chat(
+                    model=self.model,
+                    messages=[
+                        Messages(role=message.role, content=message.content)
+                        for message in messages
+                    ],
+                )
+            )
+        return ChatResponse(
+            message=ChatMessage(
+                content=response.choices[0].message.content,
+                role="assistant",
+            ),
+        )
+
+    @llm_chat_callback()
+    def chat(
+        self,
+        messages: Sequence[ChatMessage],
+        **kwargs: Any,
+    ) -> ChatResponse:
+        """Get chat."""
+        with GigaChat(**self._gigachat_kwargs) as giga:
+            response = giga.chat(
+                Chat(
+                    model=self.model,
+                    messages=[
+                        Messages(role=message.role, content=message.content)
+                        for message in messages
+                    ],
+                )
+            )
+        return ChatResponse(
+            message=ChatMessage(
+                content=response.choices[0].message.content,
+                role="assistant",
+            ),
+        )
+
+    # @llm_completion_callback()
```

Comment:
any reason to leave this commented out?

Reply:
Refactored methods, now decorator is uncommented and it works